lib/engine/game/g_22_mars/game.rb

```diff
@@ -168,18 +168,103 @@ class Game < Game::Base
           },
         ].freeze
 
+        # Game specific constants
         LAST_OR = 11
+        FIRST_REVOLT_CHECK_OR = 7
+        LAST_REVOLT_CHECK_OR = 10
 
-        @or = 0
+        def setup
+          @or = 0
+          @revolt_round = -1
+          @three_or_round = false
+        end
 
         def end_now?(_after)
           @or == LAST_OR
         end
 
+        def setup_preround
+          revolts, permits = @companies.partition(&:revolt?)
+          sorted_permits = permits.sort_by { rand }
+          sorted_revolts = revolts.sort_by { rand }
+
+          @companies = sorted_permits + sorted_revolts
```

Comment:
You're deleting most of the elements from `@companies` before you leave this method. You could just add the companies that will be given to the players:
```suggestion
          @companies = sorted_permits.take(@players.size)
          assigned_companies = sorted_permits.drop(@players.size) + sorted_revolts
```